sklearn/preprocessing/_discretization.py

```diff
@@ -225,13 +237,17 @@ def fit(self, X, y=None):
                         n_samples, size=self.subsample, replace=False
                     )
                     X = _safe_indexing(X, subsample_idx)
-        elif self.strategy != "quantile" and isinstance(
-            self.subsample, numbers.Integral
-        ):
-            raise ValueError(
-                f"Invalid parameter for `strategy`: {self.strategy}. "
-                '`subsample` must be used with `strategy="quantile"`.'
-            )
+        elif self.strategy != "quantile":
+            if isinstance(self.subsample, numbers.Integral):
+                raise ValueError(
+                    f"Invalid parameter for `strategy`: {self.strategy}. "
+                    '`subsample` must be used with `strategy="quantile"`.'
+                )
+            if sample_weight is not None:
+                raise ValueError(
+                    "`sample_weight` was provided but it can be only used with"
```

Comment:
Codecov is complaining. So you should add a test to check that we raise the error when providing a `sample_weight` and that the strategy is something else than `quantile`.

Reply:
I added `test_invalid_sample_weight` method in `test_discretization.py` (l.71) https://github.com/scikit-learn/scikit-learn/pull/22048/commits/1a417f4d1d6ee00213cd576bd41e407fc45c88e5